apps/science/src/main.cpp

```diff
@@ -1,19 +1,143 @@
+DigitalOut ledG(LED_G);
+DigitalOut ledB(LED_B);
 
-constexpr auto k_interval = 500ms;
+// Declaring can object
+CAN can(CAN1_RX, CAN1_TX, HWBRIDGE::ROVERCONFIG::ROVER_CANBUS_FREQUENCY);
 
-CANBus can(CAN1_RX, CAN1_TX, HWBRIDGE::ROVERCONFIG::ROVER_CANBUS_FREQUENCY);
+// static objects
+static mbed_error_status_t setMotionData(CANMsg &msg) {
+  float motionData;
+  msg.getPayload(motionData);
+
+  switch (msg.getID()) {
+    case HWBRIDGE::CANID::SET_GENEVA_INDEX:
+      return indexerActuator.setMotionData(motionData);
+    case HWBRIDGE::CANID::SET_SCOOPER_ANGLE:  // not sure what this case does will contact younes soon to discuss
+      return elevatorActuator.setMotionData(motionData);  // with him (perhaps this is elevatorActuator??)
+    case HWBRIDGE::CANID::SET_COVER_ANGLE:
+      return coverServo.setPosition(motionData);
+    case HWBRIDGE::CANID::SET_ELEVATOR_HEIGHT:
+      return diggerServo.setPosition(motionData);
+    default:
+      return MBED_ERROR_INVALID_ARGUMENT;
+  }
+}
+
+static CANMsg::CANMsgHandlerMap canHandleMap = {{HWBRIDGE::CANID::SET_GENEVA_INDEX, setMotionData},
+                                                {HWBRIDGE::CANID::SET_SCOOPER_ANGLE, setMotionData},
+                                                {HWBRIDGE::CANID::SET_COVER_ANGLE, setMotionData},
+                                                {HWBRIDGE::CANID::SET_ELEVATOR_HEIGHT, setMotionData}};
+
+// CAN Threads
+Thread rxCANProcessorThread;
+Thread txCANProcessorThread;
+
+// recieving CAN messages
+void rxCANProcessor() {
+  CANMsg rxMsg;
+  while (true) {
+    if (can.read(rxMsg)) {
+      canHandleMap.at(rxMsg.getID())(rxMsg);
+    }
+  }
+}
+
+// Send outgoing CAN messages
+void txCANProcessor() {
+  constexpr std::chrono::milliseconds txPeriod = 500ms;
+  CANMsg txMsg;
+
+  while (true) {
+    txMsg.setID(HWBRIDGE::CANID::REPORT_GENEVA_INDEX);
+    txMsg.setPayload(indexerActuator.getAngle_Degrees());
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
+
+    txMsg.setID(HWBRIDGE::CANID::REPORT_SCOOPER_ANGLE);
+    txMsg.setPayload(elevatorActuator.getAngle_Degrees());
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
+
+    txMsg.setID(HWBRIDGE::CANID::REPORT_COVER_ANGLE);
+    txMsg.setPayload(coverServo.read());
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
+
+    txMsg.setID(HWBRIDGE::CANID::REPORT_ELEVATOR_HEIGHT);
+    txMsg.setPayload(diggerServo.read());
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
+
+    // Read moisture returns an unsigned number so it needs to be cast to an int to be handled
+    txMsg.setID(HWBRIDGE::CANID::REPORT_MOISTURE_DATA);
+    txMsg.setPayload(static_cast<int>(moistureSensor.Read_Moisture()));
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
+
+    txMsg.setID(HWBRIDGE::CANID::REPORT_TEMPERATURE_DATA);
+    txMsg.setPayload(moistureSensor.Read_Temperature());
+    can.write(txMsg);
+    ThisThread::sleep_for(txPeriod);
+  }
```

Comment:
what you have here isnt great. we report geneva index, wait half a second, report scooper angle, wait half a second. this doesnt work bc it takes too long. the idea is we want to send all this motor data "at once" and then wait half a second and send the next batch. look at arm/src/main.cpp line 253 ish. basically u will have 2 time values one small one and one big one

Reply:
changed it to mirror the arm program